### PR TITLE
Cache bust source map fixup

### PR DESCRIPTION
**Pull request by rafde**
**What kind of change does this PR introduce?**
Feature

**Did you add tests for your changes?**
Yes

**If relevant, link to documentation update:**
Replaces https://github.com/webpack/webpack/pull/4137

**Summary**

The PR is based on a useful feature introduced by the aforementioned PR, but that PR was not updated to address PR concerns. I am trying to finish the work started.

**Does this PR introduce a breaking change?**

No breaking changes. Only augmentation.

Tried to [contact author](https://github.com/uncleyo/webpack/pull/1) of [original PR](https://github.com/webpack/webpack/pull/4137) but no response. I prefer to get this in sooner than later.

### CODE REVIEW

File: lib/SourceMapDevToolPlugin.js [hunk 1]
```diff
@@ -145,7 +146,10 @@ class SourceMapDevToolPlugin {
 							query,
 							basename: basename(filename)
 						});
-						const sourceMapUrl = path.relative(path.dirname(file), sourceMapFile).replace(/\\/g, "/");
+						let sourceMapUrl = path.relative(path.dirname(file), sourceMapFile).replace(/\\/g, "/");
+						if(sourceMapFilename.indexOf("?[contenthash]") !== -1) {
+							sourceMapUrl += "?" + crypto.createHash("md5").update(source).digest("hex");
```
Comment (sokra) [inline lib/SourceMapDevToolPlugin.js]: use `replace(/\[contenthash\]/g, ...)` instead.

File: lib/SourceMapDevToolPlugin.js [hunk 2]
```diff
@@ -145,7 +146,10 @@ class SourceMapDevToolPlugin {
 							query,
 							basename: basename(filename)
 						});
-						const sourceMapUrl = path.relative(path.dirname(file), sourceMapFile).replace(/\\/g, "/");
+						let sourceMapUrl = path.relative(path.dirname(file), sourceMapFile).replace(/\\/g, "/");
+						if(sourceMapFilename.indexOf("?[contenthash]") !== -1) {
```
Comment (sokra) [inline lib/SourceMapDevToolPlugin.js]:
```
if(sourceMapFilename.indexOf("[contenthash]") !== -1) {
```

**Comment by rafde** [comment]
Outside topic, but your bot message
> @sokra Please review the new changes.

I recommend creating a team and have the bot notify the team about new changes.

Example:
> @webpack/core Please review the new changes.

File: lib/SourceMapDevToolPlugin.js [hunk 3]
```diff
@@ -145,7 +146,10 @@ class SourceMapDevToolPlugin {
 							query,
 							basename: basename(filename)
 						});
-						const sourceMapUrl = path.relative(path.dirname(file), sourceMapFile).replace(/\\/g, "/");
+						let sourceMapUrl = path.relative(path.dirname(file), sourceMapFile).replace(/\\/g, "/");
+						if(sourceMapUrl.indexOf("[contenthash]") !== -1) {
+							sourceMapUrl = sourceMapUrl.replace("[contenthash]", crypto.createHash("md5").update(source).digest("hex"));
```
Comment (sokra) [inline lib/SourceMapDevToolPlugin.js]: Use a regexp instead to replace every occurrence.

`replace(/\[contenthash\]/g, `

File: lib/SourceMapDevToolPlugin.js [hunk 4]
```diff
@@ -5,6 +5,7 @@
 "use strict";
 
 const path = require("path");
+var crypto = require("crypto");
```
Comment (sokra) [inline lib/SourceMapDevToolPlugin.js]: const

**Comment by rafde** [comment]
CI keeps failing. I don't know if I should be concerned about some of those failing.

**Comment by sokra** [comment]
> CI keeps failing. I don't know if I should be concerned about some of those failing.

only 1 of 3 CIs fail... that's ok... one of them is always broken, that's why we have 3 of them...

**Pull request merged**